src/io/imzml/reader.rs

```diff
@@ -0,0 +1,1472 @@
+            None => Err(SpectrumAccessError::SpectrumIdNotFound(id.to_string())),
+        }
+    }
+
+    fn start_from_index(&mut self, index: usize) -> Result<&mut Self, SpectrumAccessError> {
+        match self._offset_of_index(index) {
+            Some(offset) => match self.handle.seek(SeekFrom::Start(offset)) {
+                Ok(_) => Ok(self),
+                Err(err) => Err(SpectrumAccessError::IOError(Some(err))),
+            },
+            None => Err(SpectrumAccessError::SpectrumIndexNotFound(index)),
+        }
+    }
+
+    fn start_from_time(&mut self, time: f64) -> Result<&mut Self, SpectrumAccessError> {
+        match self._offset_of_time(time) {
+            Some(offset) => match self.handle.seek(SeekFrom::Start(offset)) {
+                Ok(_) => Ok(self),
+                Err(err) => Err(SpectrumAccessError::IOError(Some(err))),
+            },
+            None => Err(SpectrumAccessError::SpectrumNotFound),
+        }
+    }
+}
+
+impl<
+        R: Read + Seek,
+        S: Read + Seek,
+        C: CentroidLike + BuildFromArrayMap,
+        D: DeconvolutedCentroidLike + BuildFromArrayMap,
+    > ImzMLReaderType<R, S, C, D>
+{
+    /// Convenience wrappers mirroring the mzML reader
+    pub fn seek(&mut self, pos: SeekFrom) -> io::Result<u64> {
+        self.handle.seek(pos)
+    }
+
+    pub fn stream_position(&mut self) -> io::Result<u64> {
+        self.handle.stream_position()
+    }
+}
+
+/// Iterator over chromatograms for imzML, paralleling the mzML implementation
+pub struct ChromatogramIter<
+    'a,
+    R: Read + Seek,
+    S: Read + Seek,
+    C: CentroidLike + BuildFromArrayMap,
+    D: DeconvolutedCentroidLike + BuildFromArrayMap,
+> {
+    reader: &'a mut ImzMLReaderType<R, S, C, D>,
+    index: usize,
+}
+
+impl<
+        'a,
+        R: Read + Seek,
+        S: Read + Seek,
+        C: CentroidLike + BuildFromArrayMap,
+        D: DeconvolutedCentroidLike + BuildFromArrayMap,
+    > ChromatogramIter<'a, R, S, C, D>
+{
+    pub fn new(reader: &'a mut ImzMLReaderType<R, S, C, D>) -> Self {
+        Self { reader, index: 0 }
+    }
+}
+
+impl<
+        R: Read + Seek,
+        S: Read + Seek,
+        C: CentroidLike + BuildFromArrayMap,
+        D: DeconvolutedCentroidLike + BuildFromArrayMap,
+    > Iterator for ChromatogramIter<'_, R, S, C, D>
+{
+    type Item = Chromatogram;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.reader.chromatogram_index.len() {
+            let result = self.reader.get_chromatogram_by_index(self.index);
+            self.index += 1;
+            result
+        } else {
+            None
+        }
+    }
+}
+
+/// For open_file we assume that the IBD file is named the same as the imzML file
+impl<C: CentroidLike + BuildFromArrayMap, D: DeconvolutedCentroidLike + BuildFromArrayMap>
+    MZFileReader<C, D, MultiLayerSpectrum<C, D>> for ImzMLReaderType<fs::File, fs::File, C, D>
+{
+    fn open_file(_source: fs::File) -> io::Result<Self> {
```

Comment:
Instead of keeping this logic in `open_file`, implement `open_path` and then use the `filename::file_name` function to infer a path and then delegate to `open_path`. Call `MzMLReaderType::build_index` on the inner mzML parser to construct the index by scanning the file there before returning since `imzML` doesn't include an `indexedmzML` wrapper.